File: vis_brain/readers/Loader.py
#-*- coding: utf-8

from vis_brain.readers import Reader


class Loader(Reader):
    '''
    Base class for all readers that support file loading
    '''

    _file = None

    def load(self, filename = ""):
        '''
        Loads the file

        filename - name of the loading file or "" is one shall use the default filename
        '''
        if filename == "":
            current_filename = self.getFilename()
        else:
            current_filename = filename
        self._file = open(current_filename, 'rb')
        matrix = self.read()
        self._file.close()
        return matrix

    def read(self):
        '''
        Reads the data from the file

        This is a virtual method that implements a certain reading algorithm based on the file structure. In this class
        the method is based however, in derived class it shall be re-implemented. In this case self._file variable
        will represent an opened file that this method shall read

        The method shall return a matrix (numpy object)
        '''
        raise NotImplementedError("vis_brain.readers.Loader.read")

File: data/reader/Loader.cpp
//
// Created by serik1987 on 12.10.2019.
//

#include "Loader.h"

namespace data::reader{

    LocalMatrix* Loader::load(mpi::Communicator& comm, const std::string &filename) {
        std::string current_filename;
        if (filename == ""){
            current_filename = getFilename();
        } else {
            current_filename = filename;
        }
        file = new mpi::File(comm, current_filename, MPI_MODE_RDONLY);
        data::LocalMatrix* result = read(comm);
        delete file;
        return result;
    }

}